### PR TITLE
RATIS-1836. Frequent retry by leader on failure path for appendEntry

### PR DESCRIPTION
**Pull request by sumitagrawl**
## What changes were proposed in this pull request?

To avoid aggresive retry when peer is unrechable failure or timeout, error wait strategy introduced

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/RATIS-1836

## How was this patch tested?

- Integration test done with ozone environment
- Unit test (pending)

### CODE REVIEW

**Review by adoroszlai — COMMENTED** [review]
Thanks @sumitagrawl for working on this.  Would it be possible to use some `RetryPolicy` (see also `RetryPolicies`) for this, instead of adding a new class with similar purpose?

**Comment by sumitagrawl** [comment]
> RetryPolicies

@adoroszlai 
- Retry policy is fixed wait retry strategy, ErrWaitStrategy can be embedded as another retry policy
- ErrWaitStrategy focus over incremental wait with failure count increasing
- GrpcLogAppender -- this is not a retry operation, but trigger events at regular interval. We can integrate behavior of RetryPolicy wait time (with ErrWaitStrategy), but may need a quite refactoring. If this is required to refactor GrpcAppender, can take as another JIRA for improvement.

**Comment by adoroszlai** [comment]
@sumitagrawl maybe I'm missing something, but `ErrWaitStrategy` very much resembles `MultipleLinearRandomRetry`

File: ratis-common/src/main/java/org/apache/ratis/util/ErrorWaitStrategy.java [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.concurrent.atomic.AtomicLong;
+
+/*
+ * Wait strategy for incremental wait based on error count.
+ *
+ */
+public class ErrorWaitStrategy {
```
Comment (szetszwo) [inline ratis-common/src/main/java/org/apache/ratis/util/ErrorWaitStrategy.java:29]: As mentioned by @adoroszlai , please use `MultipleLinearRandomRetry`.

File: ratis-grpc/src/main/java/org/apache/ratis/grpc/server/GrpcLogAppender.java [hunk 2]
```diff
@@ -101,6 +102,7 @@ public GrpcLogAppender(RaftServer.Division server, LeaderState leaderState, Foll
 
     lock = new AutoCloseableReadWriteLock(this);
     caller = LOG.isTraceEnabled()? JavaUtils.getCallerStackTraceElement(): null;
+    errorWaitStrategy = new ErrorWaitStrategy("5,1000,10,5000");
```
Comment (szetszwo) [inline ratis-grpc/src/main/java/org/apache/ratis/grpc/server/GrpcLogAppender.java:105]: The wait times should depend on `getServer().properties().minRpcTimeoutMs()`.

**Review by szetszwo — COMMENTED** [review]
@sumitagrawl , thanks for working on this.  The idea is great!  Please see the comments inlined.

**Comment by sumitagrawl** [comment]
> Thanks @sumitagrawl for working on this. Would it be possible to use some `RetryPolicy` (see also `RetryPolicies`) for this, instead of adding a new class with similar purpose?

@adoroszlai Thanks for the MultipleLinearRandomRetry class info, I have reused this for defining wait strategy reusing this.
ErrorWaitStrategy is using this to define wait duration.

File: ratis-grpc/src/main/java/org/apache/ratis/grpc/server/GrpcLogAppender.java (hunk 2 above)
Comment (sumitagrawl) [reply ratis-grpc/src/main/java/org/apache/ratis/grpc/server/GrpcLogAppender.java]: Updated...

File: ratis-common/src/main/java/org/apache/ratis/util/ErrorWaitStrategy.java (hunk 1 above)
Comment (sumitagrawl) [reply ratis-common/src/main/java/org/apache/ratis/util/ErrorWaitStrategy.java]: Updated...using this.

File: ratis-server-api/src/main/java/org/apache/ratis/server/RaftServerConfigKeys.java [hunk 3]
```diff
@@ -743,6 +743,21 @@ static TimeDuration slownessTimeout(RaftProperties properties) {
...
       setTimeDuration(properties::setTimeDuration, SLOWNESS_TIMEOUT_KEY, expiryTime);
     }
+    
+    String ERROR_WAIT_DELAY_VALUE = PREFIX + ".error.wait.delay";
+    /*
+    * min wait time as 1ms (0 is not allowed) for first 10 
+    * (5 iteration with 2 times grpc client retry)
+    * next wait 1sec for next 20 retry (10 iteration with 2 times grpc client)
+    * further wait for 5sec for max times ((5sec*980)/2 times ~= 40min)
+     */
```
Comment (szetszwo) [inline ratis-server-api/src/main/java/org/apache/ratis/server/RaftServerConfigKeys.java:753]: Use javadoc comment and align the `*`'s, i.e.
```java
/**
 *
 *
 */
```

File: ratis-server/src/main/java/org/apache/ratis/server/impl/DivisionPropertiesImpl.java [hunk 4]
```diff
@@ -28,6 +28,7 @@ class DivisionPropertiesImpl implements DivisionProperties {
   private final TimeDuration rpcTimeoutMax;
   private final TimeDuration rpcSleepTime;
   private final TimeDuration rpcSlownessTimeout;
+  private final String errorWaitDelayValue;
```
Comment (szetszwo) [inline ratis-server/src/main/java/org/apache/ratis/server/impl/DivisionPropertiesImpl.java:31]: Remove this field.  Read the property directly in `GrpcLogAppender`.

**Comment by sumitagrawl** [comment]
> @sumitagrawl , thanks for the update. Please see the comments inlined and also https://issues.apache.org/jira/secure/attachment/13058648/883_review.patch

@szetszwo Its fixed

File: ratis-server/src/test/java/org/apache/ratis/InstallSnapshotFromLeaderTests.java [hunk 5]
```diff
@@ -68,6 +68,7 @@
         prop, SNAPSHOT_TRIGGER_THRESHOLD);
     RaftServerConfigKeys.Snapshot.setAutoTriggerEnabled(prop, true);
     RaftServerConfigKeys.Log.Appender.setSnapshotChunkSizeMax(prop, SizeInBytes.ONE_KB);
+    RaftServerConfigKeys.Log.Appender.setRetryPolicy(getProperties(), "1ms,1000");
```
Comment (szetszwo) [inline ratis-server/src/test/java/org/apache/ratis/InstallSnapshotFromLeaderTests.java:71]: @sumitagrawl  Why setting 1ms in tests?  Setting it so short essentially disable the feature.  Why not using the default?

Comment (sumitagrawl) [reply ratis-server/src/test/java/org/apache/ratis/InstallSnapshotFromLeaderTests.java]: Updated, added only in one testcase where it is having wait time less for cluster to come up.

**Review by szetszwo — COMMENTED** [review]
@sumitagrawl , Thanks for the update!  Just have a minor comment inlined.

**Review by szetszwo — APPROVED** [review]
+1 the change looks good.

**Pull request merged**